dtm0/recovery.c

```diff
@@ -0,0 +1,2170 @@
+				 &dtm0_service_type,
+				 &recovery_fom_conf);
+
+		m0_sm_conf_init(&m0_drm_sm_conf);
+		rc = m0_sm_addb2_init(&m0_drm_sm_conf,
+				      M0_AVI_DRM_SM_STATE,
+				      M0_AVI_DRM_SM_COUNTER);
+
+		M0_POST(m0_sm_conf_is_initialized(&m0_drm_sm_conf));
+	}
+
+	return M0_RC(rc);
+}
+
+M0_INTERNAL void m0_drm_domain_fini(void)
+{
+	if (m0_sm_conf_is_initialized(&m0_drm_sm_conf)) {
+		m0_sm_addb2_fini(&m0_drm_sm_conf);
+		m0_sm_conf_fini(&m0_drm_sm_conf);
+		M0_POST(!m0_sm_conf_is_initialized(&m0_drm_sm_conf));
+	}
+}
+
+static void eolq_post(struct m0_dtm0_recovery_machine *m,
+		      struct eolq_item                *item)
+{
+	struct recovery_fom *rf = recovery_fom_local(m);
+	m0_be_queue_lock(&rf->rf_eolq);
+	/*
+	 * Do not post any events if we have already recovered this DTM0
+	 * service (eoq is an indicator here).
+	 */
+	if (!rf->rf_eolq.bq_the_end) {
+		/*
+		 * Assumption: the queue never gets full.
+		 * XXX: We could panic in this case. Right now, the HA/REDO FOM
+		 * should get stuck in the tick. Prorably, panic is
+		 * a better solution here.
+		 */
+		M0_BE_OP_SYNC(op, M0_BE_QUEUE_PUT(&rf->rf_eolq, &op, item));
+	}
+	m0_be_queue_unlock(&rf->rf_eolq);
+}
+
+static void heq_post(struct recovery_fom *rf, enum m0_ha_obj_state state)
+{
+	uint64_t event = state;
+	m0_be_queue_lock(&rf->rf_heq);
+	/*
+	 * Assumption: the queue never gets full.
+	 * XXX: We could panic in this case. Right now, the HA FOM should get
+	 *      stuck in the tick. Prorably, panic is a better solution
+	 *      here.
+	 */
+	M0_BE_OP_SYNC(op, M0_BE_QUEUE_PUT(&rf->rf_heq, &op, &event));
+	m0_be_queue_unlock(&rf->rf_heq);
+
+	M0_LOG(M0_DEBUG, "heq_enq " FID_F " %s ",
+	       FID_P(&rf->rf_tgt_svc),
+	       m0_ha_state2str(state));
+
+	if (!rf->rf_is_local)
+		eolq_post(rf->rf_m,
+			  &(struct eolq_item) {
+			  .ei_type = EIT_HA,
+			  .ei_ha_state = state,
+			  .ei_source = rf->rf_tgt_svc, });
+}
+
+static bool recovery_fom_ha_clink_cb(struct m0_clink *clink)
+{
+	struct recovery_fom *rf        = M0_AMB(rf, clink, rf_ha_clink);
+	struct m0_conf_obj  *proc_conf = container_of(clink->cl_chan,
+						       struct m0_conf_obj,
+						       co_ha_chan);
+	heq_post(rf, proc_conf->co_ha_state);
+	return false;
+}
+
+static int recovery_fom_init(struct recovery_fom             *rf,
+			     struct m0_dtm0_recovery_machine *m,
+			     struct m0_conf_process          *proc_conf,
+			     const struct m0_fid             *target,
+			     bool                             is_volatile)
+{
+	int rc;
+	bool is_local = m0_fid_eq(recovery_machine_local_id(m), target);
+
+	M0_ENTRY("m=%p, rf=%p, tgt=" FID_F ", is_vol=%d, is_local=%d",
+		 m, rf, FID_P(target), !!is_volatile, !!is_local);
+
+	M0_PRE(ergo(is_local, m->rm_local_rfom == NULL));
+
+	rc = m0_be_queue_init(&rf->rf_heq, &(struct m0_be_queue_cfg){
+		.bqc_q_size_max       = HEQ_MAX_LEN,
+		/* Conf-obj (1) and stop-and-wait-when-finalising (2) */
+		.bqc_producers_nr_max = 2, /* XXX */
```

Comment:
Should we remove/fix all the XXX comments?
Or what is the plan to fix them?

Reply:
XXX comments in DTM code are supposed to be fixed on future stages of development.  On current stage fixing these points is not meaningful, as we are planning changes in the design -- XXX means "in future we will know how to approach this -- and will fix at that point".

Reply:
OK

Reply:
3:Working on it